File: src/main/java/org/iota/qupla/abra/context/AbraEvalContext.java
package org.iota.qupla.abra.context;

import java.util.Arrays;
import java.util.HashMap;

import org.iota.qupla.abra.AbraBlockBranch;
import org.iota.qupla.abra.AbraBlockImport;
import org.iota.qupla.abra.AbraBlockLut;
import org.iota.qupla.abra.AbraSite;
import org.iota.qupla.abra.AbraSiteKnot;
import org.iota.qupla.abra.AbraSiteLatch;
import org.iota.qupla.abra.AbraSiteMerge;
import org.iota.qupla.abra.AbraSiteParam;
import org.iota.qupla.context.AbraContext;
import org.iota.qupla.exception.CodeException;
import org.iota.qupla.expression.FuncExpr;
import org.iota.qupla.expression.IntegerExpr;
import org.iota.qupla.expression.base.BaseExpr;
import org.iota.qupla.helper.StateValue;
import org.iota.qupla.helper.TritVector;

public class AbraEvalContext extends AbraCodeContext
{
  private static final HashMap<String, Integer> indexes = new HashMap<>();
  private static final TritVector nullTrit = new TritVector(1);
  // note: stateValues needs to be static so that state is preserved between invocations
  private static final HashMap<StateValue, StateValue> stateValues = new HashMap<>();
  public AbraContext abra;
  public int callNr;
  public byte[] callTrail = new byte[1024];
  public TritVector[] stack;
  public TritVector value;

  public AbraEvalContext()
  {
    if (indexes.size() == 0)
    {
      char key[] = new char[3];
      for (int i = 0; i < 27; i++)
      {
        key[0] = "-01".charAt((i / 1) % 3);
        key[1] = "-01".charAt((i / 3) % 3);
        key[2] = "-01".charAt((i / 9) % 3);
        indexes.put(new String(key), i);
      }
    }
  }

  public void error(final String text)
  {
    throw new CodeException(null, text);
  }

  public void eval(final AbraContext context, final BaseExpr expr)
  {
    abra = context;
    if (expr instanceof FuncExpr)
    {
      final FuncExpr funcExpr = (FuncExpr) expr;
      for (final AbraBlockBranch branch : context.abra.branches)
      {
        if (branch.origin == funcExpr.func)
        {
          value = new TritVector();
          for (final BaseExpr arg : funcExpr.args)
          {
            if (arg instanceof IntegerExpr)
            {
              value = new TritVector(value, ((IntegerExpr) arg).vector);
              continue;
            }

            error("Expected constant value");
          }

          branch.eval(this);
        }
      }
    }
  }

  @Override
  public void evalBranch(final AbraBlockBranch branch)
  {
    final TritVector[] oldStack = stack;
    stack = new TritVector[branch.siteNr];

    for (final AbraSite input : branch.inputs)
    {
      input.append(abra).newline();
      input.eval(this);
    }

    // initialize latches with old values
    for (final AbraSite latch : branch.latches)
    {
      latch.append(abra).newline();
      initializeLatch(latch);
    }

    for (final AbraSite site : branch.sites)
    {
      site.append(abra).newline();
      site.eval(this);
    }

    TritVector result = new TritVector();
    for (final AbraSite output : branch.outputs)
    {
      output.append(abra).newline();
      output.eval(this);
      result = new TritVector(result, value);
    }

    // update latches with new values
    for (final AbraSite latch : branch.latches)
    {
      latch.append(abra).newline();
      updateLatch(latch);
    }

    stack = oldStack;
    value = result;
  }

  @Override
  public void evalImport(final AbraBlockImport imp)
  {
  }

  @Override
  public void evalKnot(final AbraSiteKnot knot)
  {
    value = new TritVector();
    for (final AbraSite input : knot.inputs)
    {
      // concatenate inputs
      value = new TritVector(value, stack[input.index]);
    }

    if (value.valueTrits == 0)
    {
      stack[knot.index] = new TritVector(knot.size);
      return;
    }

    callTrail[callNr++] = (byte) knot.index;
    knot.block.eval(this);
    callNr--;

    stack[knot.index] = value;
  }

  @Override
  public void evalLatch(final AbraSiteLatch state)
  {
  }

  @Override
  public void evalLut(final AbraBlockLut lut)
  {
    if (value.trits.length() != 3)
    {
      error("LUT input size is " + value.trits.length());
    }

    if (value.valueTrits != 3)
    {
      value = nullTrit;
      return;
    }

    final Integer index = indexes.get(value.trits);
    if (index == null)
    {
      value = nullTrit;
      return;
    }

    final char trit = lut.tritCode.buffer[index];
    if (trit == '@')
    {
      value = nullTrit;
      return;
    }

    value = new TritVector("" + trit, 1);
  }

  @Override
  public void evalMerge(final AbraSiteMerge merge)
  {
    value = null;
    for (final AbraSite input : merge.inputs)
    {
      final TritVector mergeValue = stack[input.index];
      if (mergeValue.valueTrits == 0)
      {
        continue;
      }

      if (value == null)
      {
        value = mergeValue;
        continue;
      }

      error("Multiple non-null merge values");
    }

    if (value == null)
    {
      value = new TritVector(merge.size);
    }

    stack[merge.index] = value;
  }

  @Override
  public void evalParam(final AbraSiteParam param)
  {
    if (value.trits.length() < param.offset + param.size)
    {
      error("Insufficient input trits: " + value);
    }

    stack[param.index] = value.slice(param.offset, param.size);
  }

  private void initializeLatch(final AbraSite latch)
  {
    if (latch.references == 0)
    {
      return;
    }

    callTrail[callNr] = (byte) latch.index;

    final StateValue call = new StateValue();
    call.path = callTrail;
    call.pathLength = callNr + 1;

    // if state was saved before set latch to that value otherwise set to zero
    final StateValue stateValue = stateValues.get(call);
    if (stateValue != null)
    {
      stack[latch.index] = stateValue.value;
      return;
    }

    stack[latch.index] = new TritVector(TritVector.zero(latch.size), latch.size);
  }

  private void updateLatch(final AbraSite latch)
  {
    if (latch.references == 0)
    {
      return;
    }

    latch.eval(this);
    stack[latch.index] = value;

    callTrail[callNr] = (byte) latch.index;

    final StateValue call = new StateValue();
    call.path = callTrail;
    call.pathLength = callNr + 1;

    // state already saved?
    final StateValue stateValue = stateValues.get(call);
    if (stateValue != null)
    {
      // reset state?
      if (value.isZero())
      {
        stateValues.remove(call);
        return;
      }

      // overwrite state
      stateValue.value = value;
      return;
    }

    // state not saved yet

    // reset state?
    if (value.isZero())
    {
      // already reset
      return;
    }

    // save state
    call.path = Arrays.copyOf(callTrail, callNr + 1);
    call.value = value;
    stateValues.put(call, call);
  }
}

File: src/main/kotlin/org/iota/qupla/abra/context/AbraEvalContext.kt
package org.iota.qupla.abra.context

import org.iota.qupla.abra.block.AbraBlockBranch
import org.iota.qupla.abra.block.AbraBlockImport
import org.iota.qupla.abra.block.AbraBlockLut
import org.iota.qupla.abra.block.base.AbraBaseBlock
import org.iota.qupla.abra.block.site.AbraSiteKnot
import org.iota.qupla.abra.block.site.AbraSiteLatch
import org.iota.qupla.abra.block.site.AbraSiteMerge
import org.iota.qupla.abra.block.site.AbraSiteParam
import org.iota.qupla.abra.block.site.base.AbraBaseSite
import org.iota.qupla.abra.context.base.AbraBaseContext
import org.iota.qupla.helper.StateValue
import org.iota.qupla.helper.TritVector
import org.iota.qupla.qupla.context.QuplaToAbraContext
import org.iota.qupla.qupla.expression.FuncExpr
import org.iota.qupla.qupla.expression.IntegerExpr
import org.iota.qupla.qupla.expression.base.BaseExpr
import java.util.*

class AbraEvalContext : AbraBaseContext() {

    var abra: QuplaToAbraContext? = null
    var args = ArrayList<TritVector>()
    var callNr: Int = 0
    var callTrail = ByteArray(4096)
    var stack: Array<TritVector?> = Array(0) { null }
    var value: TritVector? = null

    override fun eval(context: QuplaToAbraContext, expr: BaseExpr) {
        abra = context
        if (expr is FuncExpr) {
            for (branch in context.abraModule.branches) {
                if (branch.origin === expr.func) {
                    args.clear()
                    for (arg in expr.args) {
                        if (arg is IntegerExpr) {
                            args.add(arg.vector)
                            continue
                        }

                        error("Expected constant value")
                    }

                    branch.eval(this)
                }
            }
        }
    }

    override fun evalBranch(branch: AbraBlockBranch) {
        if (branch.specialType == AbraBaseBlock.TYPE_CONSTANT) {
            value = branch.constantValue
            return
        }

        if (branch.specialType == AbraBaseBlock.TYPE_NULLIFY_TRUE) {
            if (args[0].trit(0) != '1') {
                value = branch.constantValue
                return
            }

            value = args[1]
            return
        }

        if (branch.specialType == AbraBaseBlock.TYPE_NULLIFY_FALSE) {
            if (args[0].trit(0) != '-') {
                value = branch.constantValue
                return
            }

            value = args[1]
            return
        }

        if (branch.specialType == AbraBaseBlock.TYPE_SLICE) {
            if (args.size == 1) {
                value = args[0].slice(branch.offset, branch.size)
                return
            }
        }

        val oldStack = stack
        stack = arrayOfNulls(branch.totalSites())

        if (!evalBranchInputsMatch(branch)) {
            value = null
            for (arg in args) {
                value = TritVector.concat(value, arg)
            }

            if (branch.specialType == AbraBaseBlock.TYPE_SLICE) {
                stack = oldStack
                return
            }

            for (input in branch.inputs) {
                input.eval(this)
            }
        }

        // initialize latches with old values
        for (latch in branch.latches) {
            initializeLatch(latch)
        }

        for (site in branch.sites) {
            site.eval(this)
        }

        var result: TritVector? = null
        for (output in branch.outputs) {
            output.eval(this)
            result = TritVector.concat(result, value)
        }

        // update latches with new values
        for (latch in branch.latches) {
            updateLatch(latch)
        }

        stack = oldStack
        value = result
    }

    private fun evalBranchInputsMatch(branch: AbraBlockBranch): Boolean {
        if (args.size != branch.inputs.size) {
            return false
        }

        for (i in args.indices) {
            val arg = args[i]
            val input = branch.inputs[i]
            if (arg.size() != input.size) {
                return false
            }

            stack[input.index] = arg
        }

        return true
    }

    override fun evalImport(imp: AbraBlockImport) {}

    override fun evalKnot(knot: AbraSiteKnot) {
        args.clear()
        var isAllNull = true
        for (input in knot.inputs) {
            val arg = stack[input.index]
            isAllNull = isAllNull && arg?.isNull ?: true
            args.add(arg!!)
        }

        if (isAllNull) {
            stack[knot.index] = TritVector(knot.size, '@')
            return
        }

        callTrail[callNr++] = knot.index.toByte()

        // TODO solve - not null block expected
        knot.block!!.eval(this)
        stack[knot.index] = value

        callNr--
    }

    override fun evalLatch(latch: AbraSiteLatch) {}

    override fun evalLut(lut: AbraBlockLut) {
        if (args.size != 3) {
            error("LUT needs exactly 3 inputs")
        }

        val trits = CharArray(3)
        for (i in 0..2) {
            val arg = args[i]
            if (arg.size() != 1) {
                error("LUT inputs need to be exactly 1 trit")
            }

            trits[i] = arg.trit(0)
        }

        val index = AbraBaseContext.indexFromTrits[String(trits)]
        if (index != null) {
            when (lut.lookup[index]) {
                '0' -> {
                    value = tritZero
                    return
                }

                '1' -> {
                    value = tritOne
                    return
                }

                '-' -> {
                    value = tritMin
                    return
                }
            }
        }

        value = tritNull
    }

    override fun evalMerge(merge: AbraSiteMerge) {
        value = null
        for (input in merge.inputs) {
            val mergeValue = stack[input.index]
            if (mergeValue?.isNull ?: true) {
                continue
            }

            if (value == null) {
                value = mergeValue
                continue
            }

            error("Multiple non-null merge values")
        }

        if (value == null) {
            value = TritVector(merge.size, '@')
        }

        stack[merge.index] = value
    }

    override fun evalParam(param: AbraSiteParam) {
        if (value!!.size() < param.offset + param.size) {
            error("Insufficient input trits: " + value!!)
        }

        stack[param.index] = value!!.slice(param.offset, param.size)
    }

    private fun initializeLatch(latch: AbraBaseSite) {
        if (latch.references == 0) {
            return
        }

        callTrail[callNr] = latch.index.toByte()

        val call = StateValue()
        call.path = callTrail
        call.pathLength = callNr + 1

        // if state was saved before set latch to that value otherwise set to zero
        val stateValue = stateValues[call]
        if (stateValue != null) {
            stack[latch.index] = stateValue.value
            return
        }

        stack[latch.index] = TritVector(latch.size, '0')
    }

    private fun updateLatch(latch: AbraBaseSite) {
        if (latch.references == 0) {
            return
        }

        latch.eval(this)

        // do NOT update latch site on stack with new value
        // other latches may need the old value still
        // again, do NOT add: stack[latch.index] = value;

        callTrail[callNr] = latch.index.toByte()

        val call = StateValue()
        call.path = callTrail
        call.pathLength = callNr + 1

        // state already saved?
        val stateValue = stateValues[call]
        if (stateValue != null) {
            // reset state?
            if (value!!.isZero) {
                stateValues.remove(call)
                return
            }

            // overwrite state
            stateValue.value = value
            return
        }

        // state not saved yet

        // reset state?
        if (value!!.isZero) {
            // already reset
            return
        }

        // save state
        call.path = Arrays.copyOf(callTrail, callNr + 1)
        call.value = value
        stateValues[call] = call
    }

    companion object {
        // note: stateValues needs to be static so that state is preserved between invocations
        private val stateValues = HashMap<StateValue, StateValue>()

        private val tritMin = TritVector(1, '-')
        private val tritNull = TritVector(1, '@')
        private val tritOne = TritVector(1, '1')
        private val tritZero = TritVector(1, '0')
    }
}
